language/go/fix.go

```diff
@@ -34,6 +35,110 @@ func (_ *goLang) Fix(c *config.Config, f *rule.File) {
 	removeLegacyGazelle(c, f)
 }
 
+func migrateNamingConvention(c *config.Config, f *rule.File) {
+	nc := getGoConfig(c).goNamingConvention
+
+	binName := binName(f)
+	importPath := importPath(f)
+	libName := libNameByConvention(nc, binName, importPath)
+	testName := testNameByConvention(nc, binName, importPath)
+	var migrateLibName, migrateTestName string
+	switch nc {
+	case goDefaultLibraryNC:
+		migrateLibName = libNameByConvention(importNC, binName, importPath)
+		migrateTestName = testNameByConvention(importNC, binName, importPath)
+	case importNC, importAliasNC:
+		migrateLibName = defaultLibName
+		migrateTestName = defaultTestName
+	}
+
+	for _, r := range f.Rules {
+		switch r.Kind() {
+		case "go_binary":
+			replaceInListAttr(r, "embed", ":"+migrateLibName, ":"+libName)
+		case "go_library":
+			if r.Name() == migrateLibName {
+				r.SetName(libName)
+			}
+		case "go_test":
+			if r.Name() == migrateTestName {
+				r.SetName(testName)
+				replaceInListAttr(r, "embed", ":"+migrateLibName, ":"+libName)
+			}
+		}
+	}
+
+	// Alias migration
+	if binName == "" {
+		var ar *rule.Rule
+		var lib *rule.Rule
+		for _, r := range f.Rules {
+			if r.Kind() == "alias" && r.Name() == defaultLibName {
+				ar = r
+			} else if r.Name() == libName {
+				lib = r
+			}
+		}
+		if nc == importAliasNC {
+			if ar == nil && lib != nil {
+				r := rule.NewRule("alias", defaultLibName)
+				r.SetAttr("actual", ":"+lib.Name())
+				visibility := lib.Attr("visibility")
+				if visibility != nil {
+					r.SetAttr("visibility", visibility)
+				}
+				r.Insert(f)
+			}
+		} else {
+			if ar != nil {
+				ar.Delete()
+			}
+		}
+	}
+}
+
+// binName returns the name of a go_binary rule if one can be found.
+func binName(f *rule.File) string {
+	for _, r := range f.Rules {
+		if r.Kind() == "go_binary" {
+			return r.Name()
+		}
+	}
+	return ""
+}
+
+// import path returns the existing import path from the first encountered Go rule with the attribute set.
+func importPath(f *rule.File) string {
```

Comment:
Probably better to use `InferImportPath` (from `package.go`), which is used to set `importpath` in generated rules.

Reply:
I don't see how I'm able to use InferImportPath here. I don't have any information about the Go packages by just looking at the BUILD file. What am I supposed to pass as "rel"? The file's package is no good.

Reply:
I'll withdraw my comment actually. For `migrateNamingConvention` specifically, it would be better to find the `importpath` attribute being used instead of trying to infer one based on the configuration and file location. If no `importpath` is found, it would be better to skip the transformation by returning `""` here and then returning early from `migrateNamingConvention`.

I don't think this function should be called from other transformations. They shouldn't need to support new naming conventions (for example, no one should be using `cgo_library` with the new naming convention).

For reference, the file's `Pkg` field should be the same as `rel`. Both are slash-separated paths relative to the workspace root (with `""` meaning the root itself). There may be some exotic where that's not true, but it should be fine here.

Reply:
Same response as yesterday but in the thread.

I don't think I can use `f.Pkg` because if there's a lib at the root then I can't infer its import path.

Reply:
`InfertImportPath` considers both the current prefix and `rel`, so if `rel == ""`, the import path is just the prefix. But this is probably moot: it would be better to use the current `importpath`, regardless of what we would infer.

I think the only change needed here is if we don't find any matching rules, return `""` and have `migrateNamingConvention` return early.